Add initial Spark profiler and persistence of profiling data to S3.

mbeitchman opened the pull request:
*Issue #, if available:*

https://github.com/awslabs/emr-workload-profiler/issues/2
https://github.com/awslabs/emr-workload-profiler/issues/8

*Description of changes:*

Collects Spark application metrics and save them in json in text files in S3. 

Example data: 

[ {
  "id" : "application_1513902991647_0002",
  "name" : "Spark shell",
  "attempts" : [ {
    "startTime" : "2017-12-22T21:51:52.535GMT",
    "endTime" : "2017-12-22T21:52:37.416GMT",
    "lastUpdated" : "2017-12-22T21:52:37.451GMT",
    "duration" : 44881,
    "sparkUser" : "hadoop",
    "completed" : true,
    "endTimeEpoch" : 1513979557416,
    "lastUpdatedEpoch" : 1513979557451,
    "startTimeEpoch" : 1513979512535
  } ]
}, {
  "id" : "application_1513902991647_0001",
  "name" : "Spark shell",
  "attempts" : [ {
    "startTime" : "2017-12-22T21:50:18.436GMT",
    "endTime" : "2017-12-22T21:51:06.417GMT",
    "lastUpdated" : "2017-12-22T21:51:06.459GMT",
    "duration" : 47981,
    "sparkUser" : "hadoop",
    "completed" : true,
    "endTimeEpoch" : 1513979466417,
    "lastUpdatedEpoch" : 1513979466459,
    "startTimeEpoch" : 1513979418436
  } ]
} ]


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

mbeitchman committed:
Add main entry point

This change adds the main entry point for the application and introduces
the following: logging, configuration file parsing, and creation
and execution of profilers.

This change also adds the shade plug-in to the pom file in order to
build a jar with all dependencies included. This can be built with the
mvn package command and executed with java -jar jarname.jar.

mbeitchman committed:
Add persister interface and S3 implementation

Add a persister interface and concrete implementation for pushing to
S3. The interface is created to allow for creating other types of
persisters for other desitations like local file or another data
store like Elasticsearch.

mbeitchman committed:
Add HTTP request processor

The HTTP request processor is used for making GET requests to the
application master that runs on the master node of the EMR cluster.
The application masters have endpoints for collecting metrics about
running and completed applications.

mbeitchman committed:
Add metrics collector and Spark implementation

This change adds a metrics collector interface and implementation for
Spark. The Spark metrics collector currently supports application
metrics.

Support for other metrics like job, tasks, stages, executors
and events will be added in a subsequent change.

mbeitchman committed:
Collect application metrics in Spark profiler

Add collecting of application metrics in the Spark profiler.
This change also adds persisting the metrics to S3.

mbeitchman commented on src/main/java/com/aws/emr/profiler/cli/Config.java:
```diff
@@ -0,0 +1,63 @@
+    private static Config tryParseConfig(String configPath) throws Exception {
+        Gson gson = new GsonBuilder().create();
+        File f = new File(configPath);
+        JsonReader r = new JsonReader(new InputStreamReader(new FileInputStream(f), Charset.defaultCharset()));
```
I need to close the stream reader and file input stream. Will push an update soon.

mbeitchman commented on src/main/java/com/aws/emr/profiler/core/persister/S3Persister.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Optional;
+
+public class S3Persister implements Persister {
+    private static Optional<TransferManager> transferManager;
```
Might be good to make TransferManager implement AutoClosable

mbeitchman replied:
will do this as a separate change. it will be cleaner but requires more code and this works ok

mbeitchman replied:
Will do this as a separate change.

mbeitchman commented on src/main/java/com/aws/emr/profiler/cli/Config.java:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    private static Config tryParseConfig(File configFile, Gson serializer) throws Exception {
```
Update this to use try-with resources to make sure resources are closed

mbeitchman commented:
@vinayaktAWS I updated the PR per my comments. Would love to get your feedback.

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/cli/Config.java:
```diff
@@ -0,0 +1,67 @@
+@Data
+@Builder
+public class Config {
+    private static Logger LOG = Logger.getLogger(Config.class);
```
nit: can use the @Log4j annotation from Lombok
https://projectlombok.org/api/lombok/extern/log4j/Log4j.html

mbeitchman replied:
neat, I will update the pr.

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/cli/Config.java:
```diff
@@ -0,0 +1,67 @@
+public class Config {
+    private static Logger LOG = Logger.getLogger(Config.class);
+
+    @Getter
```
Don't need these since you've got @Data

mbeitchman replied:
good catch, i will update the pr

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/cli/ProfilerFactory.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.stream.Collectors;
+
+public class ProfilerFactory {
+    private static String DEFAULT_PROFILER_CLASS ="com.aws.emr.profiler.core.";
```
Change to DEFAULT_PROFILER_PACKAGE maybe?

mbeitchman replied:
sure. good suggestion

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/cli/Config.java:
```diff
@@ -0,0 +1,67 @@
+            return Optional.of(tryParseConfig(configFile, serializer));
+        } catch (Exception ex) {
+            LOG.error(ex);
+            return Optional.empty();
```
Why are we swallowing exceptions here?

mbeitchman replied:
We really aren't swallowing the exception because we throw an `InvalidConfigFileException`  if the `Optional<Config>` is empty in the main fn per...

```java
Config config = Config.parseConfig(DEFAULT_CONFIG_PATH)
.orElseThrow(InvalidConfigFileException::new);
```

I log the original exception for context for the customer. My assumption is it will usually be a `FileNotFoundException`.

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/cli/ProfilerFactory.java:
```diff
@@ -0,0 +1,42 @@
+            Class profilerClass = Class.forName(DEFAULT_PROFILER_CLASS + profilerClassName);
+            Profiler p =  (Profiler) profilerClass.getConstructor(Config.class).newInstance(c);
+            return Optional.of(p);
+        } catch (Exception ex) {
```
Log exception?

mbeitchman replied:
sure

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/core/SparkProfiler.java:
```diff
@@ -16,9 +16,35 @@
+    private static String FILE_NAME_FORMAT = "emr-profiler-spark-metrics-%tQ";
+
+    private MetricsCollector sparkMetricsCollector;
+    private Persister s3MetricsPersister;
```
nit: rename to metricsPersister since it's the interface

mbeitchman replied:
sure

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/core/persister/S3Persister.java:
```diff
@@ -0,0 +1,82 @@
+    private void tryWriteToFile(String serializedData, File file) {
+        try {
+            Files.write(serializedData, file, Charset.forName(Charsets.UTF_8.toString()));
+        } catch (Exception ex) {
```
Log

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/core/persister/S3Persister.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Optional;
+
+public class S3Persister implements Persister {
+    private static Optional<TransferManager> transferManager;
```
Why is this an optional? Doesn't look like we are doing anything that allows an empty value

mbeitchman replied:
I use optional because I want a way to shut it down and make sure it doesn't exist. I could make it local to the function but then I have to pass it around. Ideally, transfer manager would implement the `AutoClosable` interface so it can be used in a try-with-resources statement.

The ctor with the arg for the transferManager is for testing. I like making dependency injection discrete with ctor's. I find it easier to understand than a framework.

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/core/requestprocessor/HTTPRequestProcessor.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public HTTPRequestProcessor(String host) {
+        this.INSTANCE_METADATA_URI = String.format(INSTANCE_METADATA_URI, host);
```
Is this to support external metadata servers?

mbeitchman replied:
No, ec2 supports querying this endpoint for the instance metadata. check out this link for more info: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

vinayaktAWS replied:
I meant is the second constructor there to support non 169.254.169.254 ips?

vinayaktAWS commented on src/main/java/com/aws/emr/profiler/core/requestprocessor/HTTPRequestProcessor.java:
```diff
@@ -0,0 +1,68 @@
+public class HTTPRequestProcessor {
+    private static OkHttpClient httpClient = new OkHttpClient();
+    private static String INSTANCE_METADATA_ENDPOINT = "169.254.169.254";
+    private static String INSTANCE_METADATA_URI = "http://%s/latest/meta-data/local-hostname";
```
nit: This won't be correct if the cluster is using a custom DNS server as it's provided by ec2 control plane (default VPC DNS server). Might want to use InetAddress or 'ip' commands.

Can take up support for custom dns as a different change

mbeitchman replied:
Create https://github.com/awslabs/emr-workload-profiler/issues/17 to track this for later

vinayaktAWS commented on src/resources/config.json:
```diff
@@ -0,0 +1,4 @@
+{
```
Is this a testing artifact or just a sample file?

mbeitchman replied:
just a sample file. presto provides sample configs which is very nice for getting started.

vinayaktAWS reviewed (commented):
A couple questions and a todo for custom dns

mbeitchman committed:
Use lombok log4j annotation for logger instances

This change also removes some redundant @Getter/@Setter annotations,
removes some unused imports and renames some variables to make them
more clear.

This commit address CR feedback from:

https://github.com/awslabs/emr-workload-profiler/pull/15

mbeitchman commented:
@vinayaktAWS Thank your for the awesome feedback.

I addressed all of your comments and pushed another commit to address them. Let me know if it looks good.  :sunny:

vinayaktAWS reviewed (approved):
Had a question for clarification about host as an argument for metadata endpoint in the HTTPRequestProcessor constructor but looks good!

vinayaktAWS commented:
Thank makes sense 👍  I prefer avoiding Guice/frameworks unless we think we really need them because they make often debuggability hard
